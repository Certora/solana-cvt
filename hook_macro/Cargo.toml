[package]
name = "hook-macro"
version.workspace = true
edition = "2021"
publish = false

[lib]
proc-macro = true

[dependencies]
syn = { workspace = true, features = ["full", "derive"] }
quote = { workspace = true }

[dev-dependencies]
macrotest = { workspace = true }
trybuild = { workspace = true }

