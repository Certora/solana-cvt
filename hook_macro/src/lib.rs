use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, AttributeArgs};
use quote::ToTokens;


#[proc_macro_attribute]
/**
 * This macro is used to insert a hook at the start of a function.
 * # Example
 * #[cvt_hook_start(hook())]
    fn t1() {
        // hook inserted here
        println!("t1");
    }

    expands to 

    fn t1() {
        hook();
        println!("t1");
    }
 */
pub fn cvt_hook_start(attr: TokenStream, input: TokenStream) -> TokenStream {
    
    // parse the attribute argument
    let attr = parse_macro_input!(attr as AttributeArgs);
    if attr.len() != 1 {
        panic!("Expected 1 argument");
    }

    let arg = &attr[0]; 

    // parse the input tokens and make sure it is a function
    let mut item: syn::Item = syn::parse(input).unwrap();
    let fn_item = match &mut item {
        syn::Item::Fn(fn_item) => fn_item,
        _ => panic!("Expected hook to be applied to a function")
    };


    // insert tokens_start into fn item statements at position 0
    let tokens_start = quote! { #arg; };

    fn_item.block.stmts.insert(0,syn::parse(tokens_start.into()).unwrap());

    item.into_token_stream().into()
}


#[proc_macro_attribute]
/**
 * This macro is used to insert a hook at the end of a function.
 * # Example
 * #[cvt_hook_end(hook())]
    fn t1() {
        assert_eq!(1, 1);
        // hook inserted here
        assert_eq!(2, 2);
    }

    expands to 

    fn t1() {
        assert_eq!(1, 1);
        hook();
        assert_eq!(2, 2);
    }

    Note: if the function contains only one statement,
    cvt_hook_start and cvt_hook_end are equivalent.

 */
pub fn cvt_hook_end(attr: TokenStream, input: TokenStream) -> TokenStream {

    // parse the attribute argument
    let attr = parse_macro_input!(attr as AttributeArgs);
    if attr.len() != 1 {
        panic!("Expected 1 argument");
    }

    let arg = &attr[0]; 

    // parse the input tokens and make sure it is a function
    let mut item: syn::Item = syn::parse(input).unwrap();
    let fn_item = match &mut item {
        syn::Item::Fn(fn_item) => fn_item,
        _ => panic!("Expected hook to be applied to a function")
    };

    // create tokens_end
    let tokens_end = quote! { #arg; };

    // len of fn item statements
    let len = fn_item.block.stmts.len();

    // insert tokens_end into fn item statements at position len-1
    fn_item.block.stmts.insert(len-1,syn::parse(tokens_end.into()).unwrap());
    
    item.into_token_stream().into()
}
