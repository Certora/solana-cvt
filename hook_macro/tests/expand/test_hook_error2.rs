use hook_macro::cvt_hook_start;

fn hook() {
    ();
}

fn hook2() {
    ();
}

#[cvt_hook_start(hook(), hook2())]
fn t1() {
    // hook inserted here
    println!("t1");
}

fn main() {}