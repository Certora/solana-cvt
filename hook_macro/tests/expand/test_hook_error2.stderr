error: Expected 1 argument
  --> tests/expand/test_hook_error2.rs:11:1
   |
11 | #[cvt_hook_start(hook(), hook2())]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `cvt_hook_start` (in Nightly builds, run with -Z macro-backtrace for more info)
